homeassistant/components/vicare/__init__.py

```diff
@@ -117,70 +115,72 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-    device_serial: str = device.api.getSerial()
+    device_id = device.config.getId()
+    device_serial: str | None = await hass.async_add_executor_job(
+        get_device_serial, device.api
+    )
+    device_model = device.config.getModel()
 
     old_identifier = gateway_serial
-    new_identifier = f"{gateway_serial}_{device_serial}"
+    new_identifier = (
+        f"{gateway_serial}_{device_serial if device_serial is not None else device_id}"
+    )
 
     # Migrate devices
     for device_entry in dr.async_entries_for_config_entry(registry, entry.entry_id):
-        if device_entry.identifiers == {(DOMAIN, old_identifier)}:
+        if (
+            device_entry.identifiers == {(DOMAIN, old_identifier)}
+            and device_entry.model == device_model
+        ):
             _LOGGER.debug("Migrating device %s", device_entry.name)
             registry.async_update_device(
                 device_entry.id,
                 serial_number=device_serial,
                 new_identifiers={(DOMAIN, new_identifier)},
             )
 
-
-async def async_migrate_entities(
-    hass: HomeAssistant, entry: ConfigEntry, device: ViCareDevice
-) -> None:
-    """Migrate old entry."""
-    gateway_serial: str = device.config.getConfig().serial
-    device_serial: str = device.api.getSerial()
-    new_identifier = f"{gateway_serial}_{device_serial}"
-
-    @callback
-    def _update_unique_id(
-        entity_entry: er.RegistryEntry,
-    ) -> dict[str, str] | None:
-        """Update unique ID of entity entry."""
-        if not entity_entry.unique_id.startswith(gateway_serial):
-            # belongs to other device/gateway
-            return None
-        if entity_entry.unique_id.startswith(f"{gateway_serial}_"):
-            # Already correct, nothing to do
-            return None
-
-        unique_id_parts = entity_entry.unique_id.split("-")
-        unique_id_parts[0] = new_identifier
-
-        # convert climate entity unique id from `<device_identifier>-<circuit_no>` to `<device_identifier>-heating-<circuit_no>`
-        if entity_entry.domain == DOMAIN_CLIMATE:
-            unique_id_parts[len(unique_id_parts) - 1] = (
-                f"{entity_entry.translation_key}-{unique_id_parts[len(unique_id_parts)-1]}"
-            )
-
-        entity_new_unique_id = "-".join(unique_id_parts)
-
-        _LOGGER.debug(
-            "Migrating entity %s from %s to new id %s",
-            entity_entry.entity_id,
-            entity_entry.unique_id,
-            entity_new_unique_id,
-        )
-        return {"new_unique_id": entity_new_unique_id}
-
-    # Migrate entities
-    await er.async_migrate_entries(hass, entry.entry_id, _update_unique_id)
+            @callback
+            def _update_unique_id(
+                entity_entry: er.RegistryEntry,
+            ) -> dict[str, str] | None:
+                """Update unique ID of entity entry."""
+                if entity_entry.device_id != device_entry.id:
```

Comment:
@MartinHjelmare do you have a hint for me how to fix the ruff finding here?

https://docs.astral.sh/ruff/rules/function-uses-loop-variable/